### PR TITLE
feat: Rework fieldset and new gcds-radios and gcds-checkboxes components

### PR DESCRIPTION
**Pull request by ethanWallace**
# Summary | Résumé

Updates to improve the usability of the fieldset, checkbox and radio form components. Introduces two new components, the `gcds-checkboxes` and `gcds-radios` components that will replace the previous checkbox and radio-group components. The `gcds-checkbox` and `gcds-radio-group` components will be removed in another pull request.

### Fieldset

We have updated the fieldset component so it’s focused on grouping related form fields, similar to the native `fieldset` element. The fieldset includes a legend element that provides a caption and a new `legend-size` property to align the legend size with the heading hierarchy. 

#### Overview of changes

- New `legend-size` property to set the legend’s font size to a value matching our `H2-H6` heading sizes, ensuring proper heading hierarchy and accessibility for assistive technologies.
- Removal of the `fieldset-id` property.
- Removal of the `required` property.
- Removal of the `error-message` property.
- Removal of the `disabled` property.
- Removal of the fieldset component validation.

```html
<gcds-fieldset
  legend="string"
  legend-size="h2 | h3 | h4 | h5 | h6"
  hint="string"
>
</gcds-fieldset>
```

### Radios

The `gcds-radio-group` component will be deprecated in favour of a new component `gcds-radios`. The `gcds-radios` component removes the need for the radio-group to be wrapped in a `gcds-fieldset` component. It does this by making the fieldset part of the new component and moving all validation to the new component. This change ensures better accessibility by correctly grouping radios and providing a legend. The legend font size will be equivalent to the label font size of our other GCDS form fields. This gcds-radios component will function similarly to all other GCDS form components.

#### Overview of changes

- New `gcds-radios` name
- New required `legend` property to define the legend for the built-in fieldset in the gcds-radios component.
- New `value` property to allow easy access to the value of the checked radio button.
- New `required`, `hint`, `error-message`, `disabled`, `validator` and `validate-on` properties to match the functionality of other form components in the GC Design System.

```html
<gcds-radios
  name="string"
  legend="string"
  options=’[{
    “id”: “”,
    “label”: “”,
    “value”: “”,
    “checked”: “”,
    “hint”: “”,
  }]’
  required="boolean"
  hint="string"
  error-message="string"
  disabled="booleam"
  value="string"
  validator="Array"
  Validate-on="blur || submit || other"
>
</gcds-radios>
```

### Checkboxes

The `gcds-checkbox` component will be deprecated in favour of a new component `gcds-checkboxes`. The checkboxes component will include a native `fieldset` tag that groups the checkboxes. This change removes the need for you to wrap checkboxes in the fieldset component and ensures better accessibility by correctly grouping checkboxes and providing a legend. The legend font size will be equivalent to the label font size of our other GCDS form fields.

#### Overview of changes

- New `gcds-checkboxes` name
- New `legend` property to define the legend for the built-in fieldset in the gcds-checkboxes component.
- New `options` property to pass the details for multiple checkboxes to <gcds-checkboxes>
- Rework `value` property to reflect checked checkboxes’ values.
- Removal of `label` property.
- Removal of `checkbox-id` property.
- Removal of `checked` property

```html
<gcds-checkboxes
  name="string"
  legend="string"
  options=’[{
    “id”: “”,
    “label”: “”,
    “value”?: “”,
    “checked”?: “”,
    “hint”?: “”,
  }]’
  required="boolean"
  hint="string"
  error-message="string"
  disabled="booleam"
  value="string[]"
  validator="Array"
  Validate-on="blur || submit || other"
>
</gcds-checkboxes>
```

## Notes

- Requires tokens from https://github.com/cds-snc/gcds-tokens/pull/453
- [gcds-examples branch](https://github.com/cds-snc/gcds-examples/tree/test/gcds-checkboxes) with all changes implemented in Vue and React for testing.

### CODE REVIEW

File: packages/web/src/components/gcds-radios/gcds-radios.tsx [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+   */
+  @Prop({ reflect: false, mutable: true }) errorMessage: string;
+  @Watch('errorMessage')
+  validateErrorMessage() {
+    if (this.disabled) {
+      this.errorMessage = '';
+    } else {
+      this.hasError = this.errorMessage ? !this.hasError : false;
+    }
+  }
```
Comment (daine) [inline packages/web/src/components/gcds-radios/gcds-radios.tsx:147]: This is a bit confusing, could we make this more clear? Why are we toggling `hasError` whenever `this.errorMessage` gets set or is true?

File: packages/web/src/components/gcds-radios/gcds-radios.tsx [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+   */
+  @Prop({ reflect: true, mutable: true }) value: string;
+  @Watch('value')
+  validateValue() {
+    if (this.optionsArr && this.value !== null) {
+      let isValidValue = false;
+      this.optionsArr.map(radio => {
+        if (radio.value == this.value) {
+          isValidValue = true;
+          radio.checked = true;
```
Comment (daine) [inline packages/web/src/components/gcds-radios/gcds-radios.tsx:165]: Do we need to mutate `radio.checked` here, is it possible to remove this? We're already updating if it's checked or not in this line: https://github.com/cds-snc/gcds-components/pull/845/files#diff-6cd0074b871d8bc998ce3ef3c019f185e4afcde7fab603bc53e79e0d257237f3R431

Is that not enough?

**Review by daine — CHANGES_REQUESTED** [review]
Looking really good -- I just browsed through the code and saw a few opportunities for improvement. I'll test it with the examples branch tomorrow and provide more feedback then. One thing to look out for, I'd like to test what it would be like to have multiple options with the same ID -- how it'll deal with that edge case.

**Review by melaniebmn — DISMISSED** [review]
LGTM

Things I tested in plain HTML & React:

Fieldset
- Component renders when all required props are present.
- Component doesn't render when required props are missing.
- Component properly groups form fields together.
- Component legend labels group for screen readers.
- Component passes WCAG 2.1 AA requirements.

Checkboxes
- Component renders single checkbox.
- Component renders group of checkboxes.
- Component doesn't render when required props are missing.
- Component returns correct checked value.
- Component returns error message when built-in validation doesn't pass.
- Component renders disabled, when disabled prop is present.
- Component can be used with a screen reader.
- Component passes WCAG 2.1 AA requirements.

Radios
- Component renders group of radios.
- Component doesn't render when required props are missing.
- Component returns correct checked value.
- Component returns error message when built-in validation doesn't pass.
- Component renders disabled, when disabled prop is present.
- Component can be used with a screen reader.
- Component passes WCAG 2.1 AA requirements.

File: packages/web/src/components/gcds-fieldset/gcds-fieldset.tsx [hunk 3]
```diff
@@ -36,296 +11,76 @@ export class GcdsFieldset {
...
+    const values = ['h2', 'h3', 'h4', 'h5', 'h6'];
 
-  @Watch('validator')
-  validateValidator() {
-    if (this.validator && !this.validateOn) {
-      this.validateOn = 'blur';
+    if (!values.includes(newValue)) {
+      console.error(
+        `${i18n['en'].legendSizeError} | ${i18n['fr'].legendSizeError}`,
+      );
```
[the remark this answers is not in the document]
Comment (daine) [reply packages/web/src/components/gcds-fieldset/gcds-fieldset.tsx:40]: After chatting with the devs, we're keeping it this way for now, but I'll do a follow-up conversation with the designers and the rest of the team.

**Comment by daine** [comment]
I also tested the 
[High contrast radio selection fix](https://github.com/cds-snc/gcds-components/pull/845/commits/cfb66a5171d18a0bb439ecd85625fc8f01863681) and it fixes the issue. I uploaded a video on the issue: 
- https://github.com/cds-snc/design-gc-conception/issues/1339

**Review by daine — DISMISSED** [review]
I read and tested the new commits after my last review. Even though there is a commented out test case, the risk is low since we will add it back with the new test suites. @ethanWallace could you make sure this is captured in one of the testing tickets, or stencil ticket if that's where we're putting it in?

This code looks good to me!

**Review by melaniebmn — APPROVED** [review]
LGTM
